Add radial distance checks

andkerr opened the pull request:
This PR completes work started in #825, based on changes suggested in [this](https://github.com/glotzerlab/freud/pull/825#issuecomment-913825792) comment.

## Description
I add checks for appropriate radial distance arguments in the `RDF` and `NeighborList` class constructors, and unit tests that verify that invalid parameters raise appropriate error messages, rather than produce undefined behavior.

## Motivation and Context
Resolves Issue #821 and PR #825.

Will this PR require any changes to docs for the corresponding modules?

## How Has This Been Tested?
Existing code compiles. I am sorting out the unit tests (some of my checks for error messages do not pass right now).

## Screenshots
N/A

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:
<!-- Put an `x` in all the boxes that apply. If you're unsure about any of
     these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] I have updated the documentation (if relevant).
- [x] I have added tests that cover my changes (if relevant).
- [x] All new and existing tests passed.
- [ ] I have updated the [credits](https://github.com/glotzerlab/freud/blob/master/doc/source/reference/credits.rst).
- [x] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

andkerr committed:
add additional radius/diameter checks

bdice commented on cpp/density/LocalDensity.cc:
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
+{
+    if (r_max <= 0)
+    {
+        throw std::invalid_argument("Local density calculations require that r_max must be positive");
```
This suggestion matches the style of the existing error messages more closely. See #825 for other examples.
```suggestion
        throw std::invalid_argument("LocalDensity requires r_max to be positive.");
```

bdice commented on cpp/density/LocalDensity.cc:
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
+
+    if (diameter < 0)
+    {
+        throw std::invalid_argument("Local density calculations require that diameter must be non-negative");
```
```suggestion
        throw std::invalid_argument("LocalDensity requires diameter to be non-negative.");
```

bdice commented on cpp/locality/NeighborQuery.h:
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
+    {
+        if (r_max <= 0)
+        {
+            throw std::invalid_argument("Neighbor queries require that r_max must be positive.");
```
```suggestion
            throw std::invalid_argument("NeighborQuery requires r_max to be positive.");
```

bdice commented on cpp/locality/NeighborQuery.h:
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
+        }
+        if (r_max <= r_min)
+        {
+            throw std::invalid_argument("Neighbor queries require that r_max must be greater than r_min.");
```
```suggestion
            throw std::invalid_argument("NeighborQuery requires that r_max must be greater than r_min.");
```

bdice commented on tests/test_density_LocalDensity.py:
```diff
@@ -94,3 +94,10 @@ def test_points_ne_query_points(self):
+        with pytest.raises(ValueError):
+            freud.density.LocalDensity(-1, 10)
+        with pytest.raises(ValueError):
+            freud.density.LocalDensity(2, 0)
```
Diameter is allowed to be zero. A negative value should raise an error.
```suggestion
            freud.density.LocalDensity(2, -1)
```

bdice reviewed (commented):
Thanks for working on this!

vyasr commented:
Thanks @andkerr!

andkerr committed:
Merge branch 'master' of https://github.com/glotzerlab/freud into fix/radial-distance-validation
merge in master

andkerr committed:
update error message syntax

andkerr committed:
update error message syntax

andkerr committed:
update LocalDensity invalid argument test

andkerr commented:
@bdice, I'm running into a problem that I haven't seen before. It seems like extra validation I added to the `LocalDensity` and `NeighborPointQueryIterator` constructors are causing pytest to abort on a fatal error when it tries to run the test I wrote to check that the appropriate errors are raised for invalid parameters. It looks like the problem is specifically with the additional checks in `LocalDensity.cc`. When I comment them out, the test I added simply fails with no error. 

'm not sure where my mistake is, I've done my best to match the form of other calls to `std::invalid_argument`. This is the error message I get:

![Screen Shot 2021-11-21 at 2 52 05 PM](https://user-images.githubusercontent.com/79798289/142776849-b0888bc0-bff6-49a3-92b4-8f414af782ba.png)

bdice commented:
@andkerr It's because the constructor for `LocalDensity` is missing an `except +` in the Cython bindings: https://github.com/glotzerlab/freud/blob/94e8f2404c3b8f4976634e2ce42c7eaf474dcf88/freud/_density.pxd#L38

Compare to the RDF constructor, which has `except +`:
https://github.com/glotzerlab/freud/blob/94e8f2404c3b8f4976634e2ce42c7eaf474dcf88/freud/_density.pxd#L52

Please add `except +` to all the functions (probably all constructors and `compute` methods) that can raise an error. 👍

Cython docs: https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values

andkerr committed:
add exception flag to LocalDensity Cython declaration

andkerr commented:
Great, I added the `except +` flag, tests are passing now.

bdice commented on cpp/locality/NeighborQuery.h:
```diff
@@ -265,7 +265,16 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
           m_query_point(query_point), m_finished(false), m_r_max(r_max), m_r_min(r_min),
           m_exclude_ii(exclude_ii)
-    {}
+    {
+        if (r_max <= 0)
+        {
+            throw std::invalid_argument("NeighborQuery requires r_max to be positive.");
+        }
+        if (r_max <= r_min)
+        {
+            throw std::invalid_argument("NeighborQuery requires that r_max must be greater than r_min.");
+        }
+    }
```
Can we add Python tests for this? It probably needs to be tested using concrete subclasses like AABBQuery and LinkCell. We need to be sure we get `except +` on all the methods that can construct a NeighborQuery -- that might take some careful thought.

andkerr replied:
Sure thing, I'll take care of this.

bdice replied:
@andkerr I changed the new tests a bit in ad4e6f4e9148796d7690fd749d85e812c2f2513d. It looks like you defined the tests in both AABBQuery and LinkCell classes, but the logic can be defined once in their shared parent class.

andkerr committed:
test additional r value verification

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

andkerr committed:
style fix, add docstring

andkerr committed:
Merge branch 'fix/radial-distance-validation' of github.com:glotzerlab/freud into fix/radial-distance-validation
merge in remote

andkerr committed:
unify test query calls

andkerr commented:
I added checks for appropriate errors when invalid radius arguments are passed to `AABB Query` and `LinkCell`. Both classes already had the necessary `except +` flags in the Cython declarations. I had to pass some additional argument to the calls to `query()` (`num_neighbors`, `mode`) to prevent Freud from throwing a RuntimeError, and am not sure if this was the correct choice.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

bdice commented on cpp/density/LocalDensity.cc:
```diff
@@ -12,7 +12,17 @@ namespace freud { namespace density {
+
+    if (diameter < 0)
+    {
+        throw std::invalid_argument("LocalDensity requires diameter to be non-negative");
```
```suggestion
        throw std::invalid_argument("LocalDensity requires diameter to be non-negative.");
```

bdice committed:
Use a single test in the parent NeighborQuery testing class.

bdice committed:
Update docstring.

bdice reviewed (approved):
LGTM, I pushed some small tweaks. Automerging when tests pass.

bdice committed:
Update cpp/density/LocalDensity.cc

bdice committed:
Update changelog.

bdice commented on tests/test_density_LocalDensity.py:
```diff
@@ -94,3 +94,10 @@ def test_points_ne_query_points(self):
         npt.assert_allclose(ld.density, correct_density, rtol=1e-4)
+
+    def test_invalid_radial_distances(self):
+        # make sure that invalid r_max and diameter arguments raise errrors
```
```suggestion
        """Ensure that invalid r_max and diameter arguments raise errors."""
```

bdice committed:
Update tests/test_density_LocalDensity.py

bdice committed:
Merge branch 'fix/radial-distance-validation' of github.com:glotzerlab/freud into fix/radial-distance-validation

andkerr commented:
Nice!
